projects/vdk-plugins/vdk-ipython/tests/test_plugin.py

```diff
@@ -169,3 +169,19 @@ def test_extension_with_pure_sql_job(ip, tmpdir):
     assert "2021-01-01  GOOG          123" in (
         ip.get_ipython().getoutput("! " "vdk " "sqlite-query -q 'SELECT * FROM stocks'")
     )
+
+
+def test_finalise(ip):
```

Comment:
We are missing some scenarios : 

Calling get_initialized_job_input multiple times with no issues
``
job_input = VDK.get_initialized_job_input()
job_input = VDK.get_initialized_job_input() # should be the same instance 
VDK.finalize()
```

Another scenario calling get_initialized_job_input after previously calling finalize
```
job_input = VDK.get_initialized_job_input()
VDK.finalize()
job_input_second = VDK.get_initialized_job_input()
```
 

Calling the VDK.finalize() method multiple times and verifying that it does not cause any issues.
```
job_input = VDK.get_initialized_job_input()
VDK.finalize()
VDK.finalize()
```

Scenario in if the kernel is restarted before shutdown. I don't actually know what would happen ?

Scenario in which the kernel is reset - https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-reset That removes all the variables but does not perform again the imports.

Reply:
Thanks a lot, added a  few new tests and did a few changes on the code since some of them were failing with current implementation.